### ISSUE TITLE
[css-flexbox] apply flexibility in steps

### ISSUE THREAD

**Issue by Netmosfera** [w3c/csswg-drafts#4198]
Items would be placed as normal with flexbox, but flexibility to flex items would be applied in multiple steps, where each step covers just some of the flex items, until all steps (and consequently all flex items) are covered.

Say there are 2 steps with 2 flex items each; the first step would flex such items, the second step would flex the remaining two items.

For example the first step would not fill the line entirely, maybe because of some `max-width`. Then that remaining empty space would be filled in the second step.

In plain english this would be the equivalent of:

flex-grow:
"start expanding these items **only after** these other items were expanded as much as possible"

flex-shrink:
"start shrinking these items **only after** these other items were shrank as much as possible"

In this example you can see how flex-grow is applied to red boxes first, then if there is more space available, to green, then again to blue and finally to black.

![flex](https://user-images.githubusercontent.com/4203089/63134398-a3443a00-bfc9-11e9-8550-fc8bde18ea8c.gif)

https://jsfiddle.net/ekhzrda0/1/ workaround with current flexbox features
(not very practical to use)

Thank you.

P.S. please don't abandon flexbox, I know people are super hyped up about grids, but in the real world we use flexbox just as much as grids.

**Comment by astearns** [comment]
Don't worry about Flexbox being abandoned. Flexbox and Grid have always been intended to be complementary and solve different use cases. Both are needed, and there isn't a winner or loser between them.

**Comment by SebastianZ** [comment]
For this to work, there need to be a way to define the order in which the elements grow or shrink.

And I believe this could be something that could apply to both flexbox and grid layout, maybe in form of `shrink-order` and `grow-order` properties?

So, in @Netmosfera's example you'd apply a `grow-order` of `1` to the red elements, `2` to the green ones, and `3` to the blue ones.

Sebastian

**Comment by ByteEater-pl** [comment]
I remember it used to [be in the spec](https://www.w3.org/TR/2009/WD-css3-flexbox-20090723/#propdef-box-flex-group).

**Comment by tabatkins** [comment]
You can *effectively* achieve this already by using flex values that are a few orders of magnitude different from each other. For example, for three groups you could use a 1000:1 ratio, so the first group that should grow/shrink would use values that are multiples of 1e6, the second would use 1000, and the final would use single-digit values.

**Comment by kizu** [comment]
@tabatkins Yep, that's the second example in my CodePen, and in the initial message — the main issues with this method:

1. For large items we have to use very big orders of magnitude: if we'd use just `1000`, `1000000` etc for my example, when the first items starts to shrink, the next items also shrinks _just a little_, and this leads to the very limited number of items this method supports. It is ok when we need to handle only 2-4 items, but anything more and we either cannot keep increasing the number, or get the shrinking of other items starting too soon:

    https://user-images.githubusercontent.com/177485/232154565-c8cb3fa0-1ca2-4da1-b190-ab0bdeab5060.mov

2. Transitions won't really work for this method, probably due to the way it works with the very big numbers:

    https://user-images.githubusercontent.com/177485/232154949-7f87bc99-901d-4da9-a21a-5845a24539a3.mov

CodePen demoing those two issues: https://codepen.io/kizu/pen/vYVGKRZ?editors=0100

I imagine if we'd want to implement this as something available natively, we could consider the sequential shrink/grow being for all the items between 0 and positive values or something? And the requirement would be for the transition to work nicely alongside all of these.

**Comment by kizu** [comment]
And here is _kinda_ the desired behavior for how the transition would need to work, implementing via hard-coded to the elements' widths container queries: https://codepen.io/kizu/pen/bGmpeKx

https://user-images.githubusercontent.com/177485/232160351-62669474-6336-4bf1-8b65-74b01c49fae1.mov

My main question would be: how should the regular shrink/grow interact with this? I think the simplest could be something like “any elements with explicit non-0 shrink/grow would always participate in the shrink/grow order last”.

However, I can imagine use-cases where we could want to have multiple elements shrink/grow at the same time (see the example in the top message), and in each group, we could want to use different shrink/grow values for different proportions?

**Comment by jwdallas** [comment]
A usecase for this is breadcrumbs (e.g. Apple’s Path Bar). With breadcrumbs, when space is constrained the goal is to preserve relevant information about the current position in the hierarchy as long as possible by selectively truncating less important path items in stages. If every item shrunk at the same rate to fit within the available space, none of these labels would be fully legible when space was constrained and the UI would have a lot less utility.

https://user-images.githubusercontent.com/1310725/232182947-3ba3f85c-f10e-4e5c-8eca-ce8686b8638f.mov

---
Another case to point out is the value of being able to animate flexed items without effecting the flexed size of sibling elements. If every item changed width when one item was hovered the thing you hovered would move sideways instead of out and you'd see the path bar shimmy back and forth as the cursor moved across it.

https://user-images.githubusercontent.com/1310725/232183135-9ceecd34-0df4-4f01-8bf8-e61c0672011c.mov
